q2doc/usage/usage.py

```diff
@@ -0,0 +1,200 @@
+import ast
+import functools
+import os
+import re
+from pathlib import Path
+
+from sphinx.util import logging
+
+import qiime2  # noqa: F401
+import pandas as pd  # noqa: F401
+from qiime2 import Artifact, Metadata  # noqa: F401
+import qiime2.sdk.usage as usage
+from q2cli.core.usage import CLIUsage
+from q2doc.usage.nodes import (
+    UsageNode,
+    UsageExampleNode,
+    UsageDataNode,
+    FactoryNode,
+)
+from qiime2.plugins import ArtifactAPIUsage
+
+from .utils import get_docname, get_ext
+from .validation import BlockValidator
+
+logger = logging.getLogger(__name__)
+
+
+def process_usage_blocks(app, doctree, _):
+    env = app.builder.env
+    os.chdir(env.srcdir)
+    drivers = env.domaindata['q2']['drivers']
+    for k, driver in drivers.items():
+        # ensure a new instance for each doc
+        use = driver() if isinstance(driver, type) else driver.__class__()
+        drivers[k] = use
+        processed_records = []
+        for block in env.usage_blocks:
+            tree = ast.parse(block['code'])
+            block["tree"] = tree
+            source = compile(tree, filename="<ast>", mode="exec")
+            exec(source)
+            new_records = get_new_records(use, processed_records)
+            records_to_nodes(use, new_records, block, env)
+            processed_records.extend(new_records)
+    update_nodes(doctree, env)
+
+
+def update_nodes(doctree, env):
+    exc_use = env.domaindata['q2']['drivers']['exc_use']
+    for block, node in zip(env.usage_blocks, doctree.traverse(UsageNode)):
+        nodes = block["nodes"]
+        node.replace_self(nodes)
+    for node in doctree.traverse(UsageExampleNode):
+        # Tack imports onto the first Example node.
+        node.artifact_api = node.prelude()
+        break
+    for node in doctree.traverse(FactoryNode):
+        ref = node.ref
+        try:
+            result = exc_use._get_record(ref).result
+            if isinstance(result, qiime2.metadata.metadata.Metadata):
+                metadata_preview = str(result.to_dataframe())
+                node.preview = metadata_preview
+        except KeyError as e:
+            logger.warning(
+                f"Factory `{ref}` is not used in any Usage examples: {e}"
+            )
```

Comment:
This is related to the first known issue noted here: https://3.basecamp.com/4977384/buckets/20686822/messages/3669627345